File: Python/uxadt/uxadt.py
#####################################################################
## 
## uxadt.py
##
##   A library that supports a universal, cross-platform embedded
##   representation for algebraic data type (ADT) values, and a
##   programming abstraction for operations (such as pattern
##   matching) on algebraic data type values.
##
##   Web:     uxadt.org
##   Version: 0.0.10.0
##
##

#####################################################################
## Data structure for maintaining the state of a chained matching
## expression. It must support the same matching interface as the one
## supported by values.
##

class Matching():
    candidate = None
    end = None

    # Constructor.
    def __init__(self, candidate, end = None):
        if end is None:
            self.candidate = candidate
        else:
            self.end = end

    # Matching function.
    def _(self, p, f): return self.match(p, f)
    def match(self, p, f):
        return Matching(None, self.end) if self.candidate is None else self.candidate.match(p, f)

#####################################################################
## Every UxADT value (i.e., instance of a data structure) and every
## UxADT pattern is represented as a Value object.
##

class Value():
    __ty__ = None
    __at__ = {}
    __last__ = None

    # Constructor.
    def __init__(self, entries, ty = None):
        self.__dict__.update(entries)
        self.__ty__ = ty

    # Structural equality.
    def __eq__(v, w): return v.equal(w)
    def equal(self, other):
        # If we are using stateful match storage, clean up
        # in case there is anything left over.
        self.__last__ = None

        # Compare the constructors and recursively check equality.
        for c in self.__dict__:
            for d in other.__dict__:
                if c[0:min(len(c),2)] != '__' and d[0:min(len(d),2)] != '__':
                    if c == d and len(self.__dict__[c]) == len(other.__dict__[d]):
                        for i in range(0,len(self.__dict__[c])):
                            if not self.__dict__[c][i].equal(other.__dict__[d][i]):
                                return False
                        return True
                    else:
                        return False
        return False # Failure

    # Pattern matching unification algorithm.
    @staticmethod
    def unify(p, v):
        if not isinstance(p, Value):
            return [v]

        # Compare the constructors and recursively unify if they match.
        for c in p.__dict__:
            for d in v.__dict__:
                if c[0:min(len(c),2)] != '__' and d[0:min(len(d),2)] != '__':
                    if c == d and len(p.__dict__[c]) == len(v.__dict__[d]):
                        substs = []
                        for i in range(0,len(p.__dict__[c])):
                            subst = Value.unify(p.__dict__[c][i], v.__dict__[d][i])
                            if subst is None:
                                return None
                            substs.extend(subst)
                        return substs
                    else:
                        return None
        return None

    # Matching function.
    def __lt__(self, p): return self.match(p, None)
    def _(self, p, f): return self.match(p, f)
    def match(self, p, f = None):
        # If we are using stateful match storage, clean up
        # in case there is anything left over.
        self.__last__ = None

        subst = Value.unify(p, self)
        if f == None:
            if subst != None and len(subst) > 0:
                self.__last__ = subst
            return subst != None
        else:
            return Matching(self, None if subst is None else f(*subst))

    ## Last match retrieval when using stateful matching.
    def __iter__(self): return self.matched()
    def matched(self):
        if self.__last__ != None:
            tmp = self.__last__
            self.__last__ = None
            return iter(tmp)
        raise NameError('UxADT error: can only traverse result of match exactly once after a match has occurred.')

    # Setting and getting labeled projections.
    def at(self, *args):
        if len(args) == 1:
            label = args[0]
            return self.__at__[label] if label in self.__at__ else None
        elif len(args) == 2:
            [label, proj] = args
            self.__at__[label] = proj
            return self
        return self

    # Rendering as a native nested data structure.
    def toData(self):
        for c in self.__dict__:
            if c[0:2] != '__':
                return {c: [\
                    v.toData() if type(v) == Value else v\
                    for v in self.__dict__[c]\
                    ]}

    # Rendering as a string.
    def __repr__(self): return self.toString()
    def __str__(self): return self.toString()
    def toString(self):
        for c in self.__dict__:
            if c[0:2] != '__':
                return (self.__ty__ + '.' if self.__ty__ != None else '') +\
                       c + '(' + ', '.join([str(v) for v in self.__dict__[c]]) + ')'

#####################################################################
## Functions for defining algebraic data type constructors. There
## are two ways techniques supported for introducing constructors:
##  * defining them in the local scope as stand-alone, unqualified
##    functions by passing the definition string to eval();
##  * defining a named class or object that has the named
##    constructors as its only methods.
##

def unqualified(arg1, arg2 = None):
    from types import ModuleType

    # If a name was supplied explicitly as the first argument,
    # ignore it and use the second argument for the signatures.
    sigs = arg1 if arg2 == None else arg2

    # Obtain the global context.
    import inspect
    frame = inspect.currentframe()
    context = inspect.currentframe().f_back.f_globals

    # Define the constructors as unqualified globals.
    for name in frame.f_back.f_locals:
        try:
            if type(frame.f_back.f_locals[name]) == ModuleType and frame.f_back.f_locals[name].Value == Value:
                stmts = []
                stmts += [con + " = lambda *args, **kwargs: "+name+".Value({'" + con + "': args})" for con in sigs]
                stmts = ['exec("' + s + '")' for s in stmts]
                for con in sigs:
                  context[con] = eval("lambda *args, **kwargs: Value({'" + con + "': args})")
                return '(' + ",".join(stmts) + ')'
        except: pass
    for name in frame.f_globals:
        try:
            if frame.f_globals[name] == Value:
                stmts = []
                stmts += [con + " = lambda *args, **kwargs: " + name + "({'" + con + "': args})" for con in sigs]
                stmts = ['exec("' + s + '")' for s in stmts]
                for con in sigs:
                  context[con] = eval("lambda *args, **kwargs: Value({'" + con + "': args})")
                return '(' + ",".join(stmts) + ')'
        except: pass

    # Since emitted code will refer to uxadt operations
    # by name, the module must be defined in the scope.
    # Find the module name and return the code to evaluate.
    raise NameError('UxADT error: module cannot be found in the scope. '+\
                    'Please ensure that the module is being imported correctly.')

def qualified(arg1, arg2 = None):
    # If a qualifier was supplied explicitly as the first argument,
    # use it; otherwise, make a qualifier using the names of the
    # constructors.
    if arg2 == None:
        sigs = arg1
        name = '_Type' + "_".join([con for con in sigs])
    else:
        sigs = arg2
        name = arg1

    # Create the class that has the constructors as named static methods.
    class _Type(object):
        def __init__(self, cons):
            self.__dict__.update(cons)
    cls = type(name, (_Type,), {con: eval("lambda *args: Value({'"+con+"': args}, '" + name + "')") for con in sigs}) 
    
    # Make the named class available in the global context and also return it.
    try:
        import inspect
        inspect.currentframe().f_back.f_globals[name] = cls
    except: pass
    return cls

# Synonyms for the default technique.
definition = unqualified
_ = unqualified

##eof

File: php_/uxadt.php
<?php /**************************************************************
** 
** uxadt.php
**
**   A library that supports a universal, cross-platform embedded
**   representation for algebraic data type (ADT) values, and a
**   programming abstraction for operations (such as pattern
**   matching) on algebraic data type values.
**
**   Web:     uxadt.org
**   Version: 0.0.16.0
**
*/

namespace uxadt;

/********************************************************************
** Data structure for maintaining the state of a chained matching
** expression. It must support the same matching interface as the
** one supported by values.
*/

class Matching {
  public $candidate = null;
  public $end = null;

  // Constructor.
  public function __construct($candidate, $end = null) {
    if ($end == null)
      $this->candidate = $candidate;
    else
      $this->end = $end;
  }

  // Matching function.
  public function _($p, $f) { return $this->match($p, $f); }
  public function match($p, $f) {
    return ($this->candidate == null) ? new Matching(null, $this->end) : $this->candidate->match($p, $f);
  }
}

/********************************************************************
** Every UxADT value (i.e., instance of a data structure) and every
** UxADT pattern is represented as a Value object.
*/

class Value {
  private $__ty__ = null;
  private $__at__ = array();
  private $__last__ = null;

  // Constructor.
  public function __construct($sigs, $ty = null) {
    foreach ($sigs as $con => $args)
      $this->$con = $args;
    $this->__ty__ = $ty;
  }

  // Structural equality.
  function equal($that) {
    // If we are using stateful match storage, clean up
    // in case there is anything left over.
    $this->__last__ = null;

    // Compare the constructors and recursively check equality.
    foreach (get_object_vars($this) as $c => $cc) {
      foreach (get_object_vars($that) as $d => $dd) {
        if ($c[0]!='_' && (strlen($c)<2 || $c[1]!='_') && $d[0]!='_' && (strlen($d)<2 || $d[1]!='_')) {
          if ($c == $d && count($cc) == count($dd)) {
            for ($i = 0; $i < count($cc); $i++)
              if (!($cc[$i]->equal($dd[$i])))
                return 0;
            return 1;
          } else {
            return 0;
          }
        }
      }
    }
    return 0; // Failure.
  }

  // Pattern matching unification algorithm.
  static function unify($p, $v) {
    if (!($p instanceof Value) || $p == null)
      return array($v);

    // Compare the constructors and recursively unify if they match.
    foreach (get_object_vars($p) as $c => $cc) {
      foreach (get_object_vars($v) as $d => $dd) {
        if ($c[0]!='_' && (strlen($c)<2 || $c[1]!='_') && $d[0]!='_' && (strlen($d)<2 || $d[1]!='_')) {
          if ($c == $d && count($cc) == count($dd)) {
            $substs = array();
            for ($i = 0; $i < count($cc); $i++) {
              $subst = Value::unify($cc[$i], $dd[$i]);
              if (!is_array($subst))
                return null;
              $substs = array_merge($substs, $subst);
            }
            return $substs;
          } else {
            return null;
          }
        }
      }
    }
    return null;
  }

  // Matching function.
  function _($p, $f = null) { return $this->match($p, $f); }
  function match($p, $f = null) {
    // If we are using stateful match storage, clean up
    // in case there is anything left over.
    $this->__last__ = null;

    $subst = Value::unify($p, $this);
    if ($f == null) {
      if (is_array($subst) && count($subst) > 0)
        $this->__last__ = $subst;
      return is_array($subst);
    } else {
      return new Matching($this, (!is_array($subst)) ? null : call_user_func_array($f, $subst));
    }
  }

  // Last match retrieval when using stateful matching.
  function __invoke() { return $this->matched(); }
  function matched() {
    if ($this->__last__ != null) {
      $tmp = $this->__last__;
      $this->__last__ = null;
      return $tmp;
    }
    throw new ErrorException('UxADT error: can only traverse result of match exactly once after a match has occurred.');
  }

  // Setting and getting labeled projections.
  function at () {
    $args = func_get_args();
    if (count($args) == 1) {
      $label = $args[0];
      return array_key_exists($label, $this->__at__) ? $this->__at__[$label] : null;
    } else if (count($args) == 2) {
      list($label, $proj) = $args;
      $this->__at__[$label] = $proj;
      return $this;
    }
    return $this;
  }

  // Rendering as a native nested data structure.
  public function toData() {
    $s = "";
    $ss = array();
    foreach (get_object_vars($this) as $c => $cc) {
      if ($c[0] != '_' && (strlen($c) < 2 || $c[1] != '_')) {
        for ($i = 0; $i < count($cc); $i++)
          array_push($ss, (is_a($cc[$i], '\uxadt\Value')) ? $cc[$i]->toData() : $cc[$i]);
        return array($c => $ss);
      }
    }
  }

  // Rendering as a string.
  static private function arrayToStringRecursive($xs) {
    // Handle arrays by emitting explicit "array()" calls.
    // TODO: Handle strings by escaping them properly and
    // emitting quotation delimiters.
    $ss = array();
    foreach ($xs as $x) {
      if (is_a($x, '\uxadt\Value'))
        array_push($ss, $x->toString());
      else if (is_string($x)) 
        array_push($ss, "'" . addslashes($x) . "'");
      else if (is_array($x))
        array_push($ss, arrayToStringRecursive($x));
      else
        array_push($ss, ''.$x);
    }
    return 'array(' . implode(", ", $ss) . ')';
  }
  public function __toString() { return $this->toString(); }
  public function toString() {
    $s = "";
    $ss = array();
    foreach (get_object_vars($this) as $c => $cc) {
      if ($c[0] != '_' && (strlen($c) < 2 || $c[1] != '_')) {
        $s = (($this->__ty__ != null) ? $this->__ty__ . '::' : '') . $c . '(';
        foreach ($cc as $v) {
          if (is_a($v, '\uxadt\Value'))
            array_push($ss, $v->toString());
          else if (is_string($v)) 
            array_push($ss, "'" . addslashes($v) . "'");
          else if (is_array($v))
            array_push($ss, Value::arrayToStringRecursive($v));
          else
            array_push($ss, ''.$v);
        }
        return $s . implode(", ", $ss) . ')';
      }
    }
  }
}

/********************************************************************
** Functions for defining algebraic data type constructors. There
** are two ways techniques supported for introducing constructors:
**  * defining them in the local scope as stand-alone, unqualified
**    functions by passing the definition string to eval();
**  * defining a named class or object that has the named 
**    constructors as its only methods.
*/

function unqualified($arg1, $arg2 = null) {
  // If a name was supplied explicitly as the first argument,
  // ignore it and use the second argument for the signatures.
  $sigs = ($arg2 === null) ? $arg1 : $arg2;

  // Since emitted code will refer to UxADT operations
  // by name, the module must be defined in the scope.
  $defs =
      "if (!(class_exists('\\uxadt\\Value')))"
    . "throw new ErrorException('"
    . "UxADT error: module cannot be found in the scope."
    . "Please ensure that the module is being imported correctly."
    . "');";

  // Define the constructors as unqualified globals.
  foreach ($sigs as $con => $args)
    $defs .= 'function '.$con.' () { $v = new \\uxadt\\Value(array("'.$con.'" => func_get_args())); return $v; };';
  eval($defs);
}

function qualified($arg1, $arg2 = null, $arg3 = null) {
  // There are three supported possibilities for supplied arguments:
  // * (<constructors>)
  // * (<qualifier>, <constructors>)
  // * (<namespace>, <qualifier>, <constructors>)
  // We handle each case separately.
  if ($arg2 === null && $arg3 === null) {
    $sigs = $arg1;
    $name = "_Type";
    $ns = "";
    foreach ($sigs as $con => $args)
      $name .= "_" . $con;
  } else if ($arg3 === null) {
    $sigs = $arg2;
    $name = $arg1;
    $ns = "";
  } else {
    $sigs = $arg3;
    $name = $arg2;
    $ns = $arg1;
  }

  $ns_decl = ($ns !== "") ? "namespace " . $ns . ";" : "";
  $ns = ($ns !== "") ? "\\" . $ns . "\\" : "";

  // Create the class that has the constructors as named static methods.
  $defs = $ns_decl . "class " . $name . "{ ";
  foreach ($sigs as $con => $args)
    $defs .= "public static function " . $con . "() { return new \\uxadt\\Value(array('" . $con . "' => func_get_args()), '" . $ns . $name . "'); } ";
  $defs .= ' };';

  // Make the named object of that class available in the global context and also return it.
  eval($defs);
  eval('$t = new ' . $ns . $name . '();');
  return $t;
}

// Synonyms for the default technique.
function definition($arg1, $arg2) { return unqualified($arg1, $arg2); }
function _($arg1, $arg2 = null) { return unqualified($arg1, $arg2); }

/*eof*/ ?>